### PR TITLE
Add interface to writer modules

### PR DESCRIPTION
**Pull request by Spacetown**
This is the second approach to create a interface for the different formats because #474 was outdated.

As suggested in the previous PR the reader and writer code is now loaded on demand which can improve the startup performance of `gcovr`. Also the `gcov` format is now handled as a reader and we are open to extend to other input formats.

The whole handling of the exit codes is done in `__main__.py`. Constants for the exit codes are added. A separate exitcode for the reading code is introduced which results in a different exit code if th option `--add-tracefile` doesn't match a file (old 1, new 8)

The options are prepared for each format separate to give only the common options and the format specific options to the reader/writer.

### CODE REVIEW

File: CHANGELOG.rst [hunk 1]
```diff
@@ -23,6 +23,7 @@ Documentation:
 Internal changes:
 
 - Do not scrub versions in reference data. (:issue:`747`)
+- Add interface for the different formats to easily add new formats. (:issue:`755`)
```
Comment (latk) [inline CHANGELOG.rst:26]: The Changelog does not yet mention the new CLI option aliases like `--txt-summary`.

**Review by latk — APPROVED** [review]
Thank you for doing this work! I think this refactoring helps make the individual formats more manageable, and editing command line options should now be less scary than having to scroll through a 1000-line file.

I think most formats don't need to be split into separate `__init__` and `write` files. The ability to lazily load implementations is most relevant for the XML and HTML formats, and we now have a very high degree of namespace nesting (`from ...utils import foo` is not necessarily obvious). However, I don't feel strongly about this.

I have stronger opinions about the design of the handler classes, which I've commented inline.

If there is any doubt about how to proceed, I would suggest to rather merge this as-is, and fix any issues in later PRs. All of these changes should have purely internal effects and are reversible if needed (aside from the new CLI option aliases). This is clearly a step in a good direction.

File: gcovr/formats/base.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+from ..options import GcovrConfigOption, Options
+from ..coverage import CovData
+
+
+class handler_base(object):
+    def get_options() -> List[GcovrConfigOption]:
+        return []
+
+    @staticmethod
+    def read_report(covdata: CovData, options: Options) -> bool:
```
[the remark this answers is not in the document]
Comment (Spacetown) [reply gcovr/formats/base.py:31]: The writer already used this error code (introduced with #358). I'm with you that a exception handling would be better here.

**Comment by Spacetown** [comment]
> Thank you for doing this work! I think this refactoring helps make the individual formats more manageable, and editing command line options should now be less scary than having to scroll through a 1000-line file.

It was also a long wish from my side.

> I think most formats don't need to be split into separate `__init__` and `write` files. The ability to lazily load implementations is most relevant for the XML and HTML formats, and we now have a very high degree of namespace nesting (`from ...utils import foo` is not necessarily obvious). However, I don't feel strongly about this.

I've done this to have the same structure for each format. How is about json formats, is the loading of the JSON stuff faster than the XML once?

> I have stronger opinions about the design of the handler classes, which I've commented inline. 
> If there is any doubt about how to proceed, I would suggest to rather merge this as-is, and fix any issues in later PRs. All of these changes should have purely internal effects and are reversible if needed (aside from the new CLI option aliases). This is clearly a step in a good direction.

I'll push new commits for your suggestions.

File: gcovr/formats/base.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+#
+# ****************************************************************************
+
+from typing import List
+
+from ..options import GcovrConfigOption, Options
+from ..coverage import CovData
+
+
+class handler_base(object):
```
[the remark this answers is not in the document]
Comment (Spacetown) [reply gcovr/formats/base.py:26]: I've added a constructor which extracts the options.

File: gcovr/options.py [hunk 4]
```diff
@@ -295,8 +295,13 @@ def __init__(
...
             help += " Negation: {}.".format(", ".join(negate))
-        if (flags or positional) and config_key and "--" + config_key not in flags:
-            help += f" Config key: {config_key}."
+        if (flags or positional) and config_keys:
+            for config_key in config_keys:
+                config_keys_help = []
+                if "--" + config_key not in flags:
+                    config_keys_help.append(config_key)
+                if len(config_keys_help) > 0:
+                    help += f" Config key(s): {', '.join(config_keys_help)}."
```
[the remark this answers is not in the document]
Comment (Spacetown) [reply gcovr/options.py:304]: Argh, I missed this. Would it be better if we always add the config keys to the help? E.g. if someone specifies a single key which is matching one of the long options nothing will be printed. It looks the same as the default where all long options are useable as config keys.

[the remark this answers is not in the document]
Comment (Spacetown) [reply gcovr/options.py:304]: Raised #761 for fixing this.

**Pull request merged**